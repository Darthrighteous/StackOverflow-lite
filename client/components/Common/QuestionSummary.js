import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import moment from 'moment';
import getUserDetails from '../../utils/getUserDetails';

/**
 * Question summary component
 */
class QuestionSummary extends Component {
  constructor(props) {
    super(props);
    const questionPath = `/question/${props.question.id}`;
    this.navigateToQuestion = this.navigateToQuestion.bind(this, questionPath);
  }

  stopPropagation(e) {
    e.stopPropagation();
  }
  
  navigateToQuestion(path) {
    const { history } = this.props;
    history.push(path);
  }

  render() {
    const { question } = this.props;
    const { answer_count, score, title, username, id, created_at } = question;

    let isOwner = false;
    const loggedInUser = getUserDetails();
    if(loggedInUser) {
      isOwner = loggedInUser.username === question.username;
    }
    const questionPath = `/question/${id}`;
    return (
      <div className="question-summary" role="button" tabIndex="0" onKeyDown={null} onClick={this.navigateToQuestion}>
        <div className="question-stats">
          <div className="stat-answers">
            <span>{answer_count}</span>
            <h5>answers</h5>
          </div>
          <div className="stat-score">
            <span>{score}</span>
            <h5>score</h5>
          </div>
        </div>

        <div className="summary-body">
          <h3>{title}</h3>
          <div className="summary-details">
            <Link
              to={questionPath}
              onClick={this.stopPropagation}
              className="details-date"
            >
              {moment(created_at).fromNow()}
            </Link>
            by
            <Link
              to={`/profile/${username}`}
              onClick={this.stopPropagation}
              className="details-user"
            >
              {username}
            </Link>          
          </div>
        </div>

        <div className="summary-options">
          <Link to={`${questionPath}?focus=answer`} onClick={this.stopPropagation} id="comment-icon">
            <i className="fas fa-comment-alt" />
          </Link>
          {(isOwner) ? (
            <Link to={questionPath} onClick={this.stopPropagation} id="delete-icon">
              <i className="fas fa-trash" />
            </Link>
          ) : (
            <div />
          )}
        </div>
      </div>
    );
  }
}

QuestionSummary.propTypes = {
  question: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
};

export default QuestionSummary;