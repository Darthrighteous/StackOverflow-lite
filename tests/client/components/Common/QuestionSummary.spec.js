import React from 'react';
import { shallow } from 'enzyme';

import QuestionSummary from '../../../../client/components/Common/QuestionSummary';
import { mockQuestionsResponse } from '../../../support/mockData';

/**
 * setup for test
 * @returns {object} containing props
 */
function setup() {
  const props = {
    history: { push: jest.fn() },
    question: mockQuestionsResponse.questions[0],
  };

  const questionSummary = shallow(
    <QuestionSummary {...props} isOwner={false} />
  );
  const ownQuestionSummary = shallow(
    <QuestionSummary {...props} isOwner />
  );

  return {
    props,
    questionSummary,
    ownQuestionSummary,
  };
}

describe('Question summary snapshot tests', () => {
  const { questionSummary, ownQuestionSummary } = setup();

  test('for non owner', () => {
    expect(questionSummary).toMatchSnapshot();
  });

  test('for owner', () => {
    expect(ownQuestionSummary).toMatchSnapshot();
  });
});

describe('event simulations', () => {
  const { questionSummary } = setup();
  it('should stop propagation of child element clicks', () => {
    const event = {
      stopPropagation: jest.fn()
    };
    questionSummary.find('#comment-icon').simulate('click', event);
    expect(event.stopPropagation).toHaveBeenCalled();
  });
});

describe('handler unit tests', () => {
  const { props, questionSummary } = setup();
  const instance = questionSummary.instance();
  it('should navigate', () => {
    instance.navigateToQuestion('/question/1');
    expect(props.history.push).toHaveBeenCalled();
  });
});
