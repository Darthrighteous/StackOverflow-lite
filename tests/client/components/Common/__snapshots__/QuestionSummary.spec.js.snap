// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question summary snapshot tests for non owner 1`] = `
<div
  className="question-summary"
  onClick={[Function]}
  onKeyDown={null}
  role="button"
  tabIndex="0"
>
  <div
    className="question-stats"
  >
    <div
      className="stat-answers"
    >
      <span>
        0
      </span>
      <h5>
        answers
      </h5>
    </div>
    <div
      className="stat-score"
    >
      <span>
        0
      </span>
      <h5>
        score
      </h5>
    </div>
  </div>
  <div
    className="summary-body"
  >
    <h3>
      Why do we celebrate when we win?
    </h3>
    <div
      className="summary-details"
    >
      <Link
        className="details-date"
        onClick={[Function]}
        replace={false}
        to="/question/4"
      >
        2 days ago
      </Link>
      by
      <Link
        className="details-user"
        onClick={[Function]}
        replace={false}
        to="/user/Darthrighteous"
      >
        Darthrighteous
      </Link>
    </div>
  </div>
  <div
    className="summary-options"
  >
    <Link
      id="comment-icon"
      onClick={[Function]}
      replace={false}
      to="/question/4?focus=answer"
    >
      <i
        className="fas fa-comment-alt"
      />
    </Link>
    <div />
  </div>
</div>
`;

exports[`Question summary snapshot tests for owner 1`] = `
<div
  className="question-summary"
  onClick={[Function]}
  onKeyDown={null}
  role="button"
  tabIndex="0"
>
  <div
    className="question-stats"
  >
    <div
      className="stat-answers"
    >
      <span>
        0
      </span>
      <h5>
        answers
      </h5>
    </div>
    <div
      className="stat-score"
    >
      <span>
        0
      </span>
      <h5>
        score
      </h5>
    </div>
  </div>
  <div
    className="summary-body"
  >
    <h3>
      Why do we celebrate when we win?
    </h3>
    <div
      className="summary-details"
    >
      <Link
        className="details-date"
        onClick={[Function]}
        replace={false}
        to="/question/4"
      >
        2 days ago
      </Link>
      by
      <Link
        className="details-user"
        onClick={[Function]}
        replace={false}
        to="/user/Darthrighteous"
      >
        Darthrighteous
      </Link>
    </div>
  </div>
  <div
    className="summary-options"
  >
    <Link
      id="comment-icon"
      onClick={[Function]}
      replace={false}
      to="/question/4?focus=answer"
    >
      <i
        className="fas fa-comment-alt"
      />
    </Link>
    <Link
      id="delete-icon"
      onClick={[Function]}
      replace={false}
      to="/question/4"
    >
      <i
        className="fas fa-trash"
      />
    </Link>
  </div>
</div>
`;
